include/tvm/relay/attrs/memory.h

```diff
@@ -27,10 +27,37 @@
 #include <tvm/ir/attrs.h>
 #include <tvm/relay/expr.h>
 #include <string>
+#include <vector>
 
 namespace tvm {
 namespace relay {
 
+std::vector<TensorType> FlattenTupleType(const Type& type);
+std::vector<Expr> FromTupleType(const Type& type, const Expr& expr);
+Expr ToTupleType(const Type& t, const Array<Expr>& exprs);
+
+/*!
+ * \brief Options for allocating storage.
+ */
+struct AllocStorageAttrs : public tvm::AttrsNode<AllocStorageAttrs> {
+  DataType dtype;
+  int device_id;
+  int device_type;
+
+  TVM_DECLARE_ATTRS(AllocStorageAttrs, "relay.attrs.AllocStorageAttrs") {
+    TVM_ATTR_FIELD(dtype)
+      .describe(
+         "The dtype of the tensor to allocate.")
+      .set_default(DataType::Float(32, 1));
+    TVM_ATTR_FIELD(device_id)
```

Comment:
We should not add `device_id` in the attribute. Instead, we should allow user to specify the device id in the ctx and give it to vm when initialize it.

Reply:
@tqchen and I talked about support for mapping virtual contexts to logical contexts, i.e schedule for n devices but map from 1 to n at runtime. This allows us to do a mixture of static and dynamic scheduling. Now that I think about it we probably want to logically group the allocations based on device id either way to ensure we place them on the right device (i.e which device do I place an allocation when I have n-gpus). 

I want to store the the context directly but it doesn't extend Object, i.e is not a valid attribute type.

Reply:
I think both of you talks about the same thing, but perhaps need some clarifications.

I think the `device_id` here actually means a virtual device id that corresponds to a virtual device, and perhaps should not be confused with the actual device.

Some alternative related APIs are


- S0 `allocate_storage(device_type, TranslateDeviceIndex(virtual_device_id))`: directly calls low level allocation, translate device id translate virtual one into the real device
- S1 `allocate_storage(virtual_device_id)`: implicitly calls translation under the hood

Reply:
I now understand that you want to have a general alloc storage instruction that can support multiple same type devices. If it's a virtual device, probably we can go with S1 since we can map the `virtual_device_id` to a tuple of `device_type` and physical `device_id`. 

I also wonder how you plan to design the user-facing interface that would use multiple same type devices. By user annotation? Or just specifying number of devices and the memory planning does the allocation?

Reply:
@icemelon9 One option Jared and I talked before is probably to leverage the `on_device` annotation first. I think we can start from user annotation.

Reply:
Okay let's talk to today during our meeting about the design. I'm happy to iterate

Reply:
After a discussion with Jared, we agreed that we can keep the device id for future proof to support multiple devices.